### ISSUE TITLE
Cannot alias keywords mypy has special cases for

### ISSUE THREAD

**Issue by sshane** [python/mypy#17656]
**Bug Report**

If you want to alias or wrap a keyword, mypy cannot infer the original keyword.

**To Reproduce**

Take this example, this works:

```python
var: int | str = 1

if isinstance(var, int):
  print(var + 1)
```

but this raises an error:

```python
var: int | str = 1

which = isinstance

if which(var, int):
  print(var + 1)
```

```
batman@workstation-shane:~/openpilot$ mypy selfdrive/car/data_structures.py 
selfdrive/car/data_structures.py:265: error: Unsupported operand types for + ("str" and "int")  [operator]
selfdrive/car/data_structures.py:265: note: Left operand is of type "int | str"
Found 1 error in 1 file (checked 1 source file)
```

---

Another case is wrapping dataclass to implement custom pre-processing/other behavior:

```python
def custom_dataclass(cls, /, **kwargs):
  # do some work on cls
  return dataclass(cls, **kwargs)


# or even `custom_dataclass = dataclass`

@custom_dataclass
class MyClass:
  test: int = 0


MyClass(1)
```

```
batman@workstation-shane:~/openpilot$ mypy selfdrive/car/data_structures.py 
selfdrive/car/data_structures.py:278: error: Too many arguments for "MyClass"  [call-arg]
Found 1 error in 1 file (checked 1 source file)
```

However, `from dataclasses import dataclass as _dataclass` works.

**Expected Behavior**

Mypy can deduce what the alias actually is, or what the wrapper contains inside it.

**Actual Behavior**

See above.

**Your Environment**

mypy 1.11.1 (compiled: yes)
Python 3.11.4

**Comment by erictraut** [comment]
These two examples are very different. The first defines a direct alias, and I would expect that to work.

The second is not an alias, but a wrapper. From a static perspective, there's no association between `custom_dataclass` and `dataclass`. If you want this to work with a static type checker, you need to provide it with some additional information. In particular, you need to decorate your `custom_dataclass` with `typing.dataclass_transform`.

```python
from dataclasses import dataclass, field
from typing import dataclass_transform

@dataclass_transform(field_specifiers=(field,))
def custom_dataclass(cls, /, **kwargs):
    return dataclass(cls, **kwargs)
```

**Comment by sshane** [comment]
I did not know about `dataclass_transform`, thanks for the tip! Yes, then that leaves the direct aliasing as the issue here.

**Comment by sshane** [comment]
I didn't have a use for the `which = isinstance` example, so going to close as `dataclass_transform` solves my main issue. Thanks again!
